### PR TITLE
#201 Replace TSLint by ESLint

### PR DESCRIPTION
**Pull request by agmitron**
[#201 ](https://github.com/Haulmont/jmix-frontend/issues/201) Replace TSLint by ESLint

### CODE REVIEW

**Comment by vyacheslav-pushkin** [comment]
Please check why CI is failing.

**Comment by vyacheslav-pushkin** [comment]
`eslint` and `@typescript-eslint/eslint-plugin` are missing in dev deps, at least in `react-ui` (didn't check others)

**Comment by vyacheslav-pushkin** [comment]
I vote to disable some rules:
- indentation rule as it seems to generate a lot of false positives. 
- `Unnecessary { after 'if' condition ` we do require the curly brace after if / else / for of / etc.
- ` Type boolean trivially inferred from a boolean literal, remove type annotation` - it complains about stuff like `disabled: boolean = false` in function signature
- `@typescript-eslint/no-empty-interface` - sometimes we might need a simple alias for a complex interface, or use our own interface instead of external one so that we can extend it in future

These constitute the bulk of the errors/warnings, the rest should be fairly trivial to fix, which is what I suggest to do as part of this PR.

@web-devel WDYT?

File: packages/jmix-react-core/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      2,
+      {
+        "allow": [
+          "warn",
+          "error"
+        ]
+      }
+    ],
+    "react-hooks/rules-of-hooks": "error",
+    "react-hooks/exhaustive-deps": "warn",
```
Comment (vyacheslav-pushkin) [inline packages/jmix-react-core/.eslintrc.json:61]: I think should be an `error`. In super-rare cases when we don't want exhaustive deps we can turn off the rule with a comment.

File: packages/jmix-react-core/.eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      "warn",
+      {
+        "argsIgnorePattern": "^_"
+      }
+    ],
+    "@typescript-eslint/no-non-null-assertion": "off",
+    "@typescript-eslint/ban-ts-comment": "off",
+    "react/display-name": "off",
+    "no-prototype-builtins": "warn",
+    "no-fallthrough": "off",
```
Comment (vyacheslav-pushkin) [inline packages/jmix-react-core/.eslintrc.json:76]: Let' turn it on.

**Review by vyacheslav-pushkin — CHANGES_REQUESTED** [review]

**Comment by vyacheslav-pushkin** [comment]
Also my IDEA doesn't seem to pick up eslint's warnings/errors.

**Comment by web-devel** [comment]
> @web-devel WDYT?
Agree with all points.

File: packages/jmix-react-core/.eslintrc.json (hunk 1 above)
Comment (agmitron) [reply packages/jmix-react-core/.eslintrc.json]: **Resolved**: https://github.com/Haulmont/jmix-frontend/pull/311/commits/a21541fede9704eb43aae90e73d940ceb540367c

File: packages/jmix-react-core/.eslintrc.json (hunk 2 above)
Comment (agmitron) [reply packages/jmix-react-core/.eslintrc.json:73]: **Resolved**: https://github.com/Haulmont/jmix-frontend/pull/311/commits/a21541fede9704eb43aae90e73d940ceb540367c

File: packages/jmix-react-core/src/app/Router.tsx [hunk 3]
```diff
@@ -34,11 +34,13 @@ function exec(re: RegExp, str: string, keys: pathToRegexp.Key[]) {
...
-export function redirect(to: string, replace = false, title = '') {
-  const currentFullPath = window.location.href.substr(window.location.origin.length);
+export function redirect(to: string, replace = false, title = "") {
+  const currentFullPath = window.location.href.substr(
+    window.location.origin.length
+  );
   if (currentFullPath === to) return;
 
-  if (currentRoute.hashMode) to = '#' + to;
+  if (currentRoute.hashMode) to = "#" + to;
```
Comment (vyacheslav-pushkin) [inline packages/jmix-react-core/src/app/Router.tsx:43]: I suggest either to enforce single quote, or to allow both single and double quote.
@web-devel WDYT?

Comment (web-devel) [reply packages/jmix-react-core/src/app/Router.tsx:43]: I vote for allowing both

**Pull request closed without merging**